File: Hammerspoon 2/Modules/hs.timer/hs.timer.js
//
//  hs.timer.js
//  Hammerspoon 2
//
//  Created by Chris Jones on 06/11/2025.
//

"use strict";

// Predicate-based timers
// These are stored as Swift-retained JSValue properties on HSTimerModule to prevent garbage collection.

/// Repeat a function/lambda until a given predicate function/lambda returns true
/// Parameters:
///  - predicateFn: A function/lambda to test if the timer should continue. Return True to end the timer, False to continue it
///  - actionFn: A function/lambda to call until the predicateFn returns true
///  - checkInterval: How often, in seconds, to call actionFn
hs.timer.doUntil = function(predicateFn, actionFn, checkInterval) {
    if (typeof predicateFn !== 'function') {
        throw new Error("hs.timer.doUntil(): predicate must be a function");
    }
    if (typeof actionFn !== 'function') {
        throw new Error("hs.timer.doUntil(): action must be a function");
    }

    checkInterval = checkInterval || 1;

    const timer = hs.timer.create(checkInterval, function() {
        if (predicateFn()) {
            actionFn();
            timer.stop();
        } else {
            actionFn();
        }
    });

    timer.start();
};

/// Repeat a function/lambda while a given predicate function/lambda returns true
/// Parameters:
///  - predicateFn: A function/lambda to test if the timer should continue. Return True to continue the timer, False to end it
///  - actionFn: A function/lambda to call while the predicateFn returns true
///  - checkInterval: How often, in seconds, to call actionFn
hs.timer.doWhile = function(predicateFn, actionFn, checkInterval) {
    if (typeof predicateFn !== 'function') {
        throw new Error("hs.timer.doWhile(): predicate must be a function");
    }
    if (typeof actionFn !== 'function') {
        throw new Error("hs.timer.doWhile(): action must be a function");
    }

    checkInterval = checkInterval || 1;

    const timer = hs.timer.create(checkInterval, function() {
        if (!predicateFn()) {
            timer.stop();
        } else {
            actionFn();
        }
    });

    timer.start();
};

/// Wait to call a function/lambda until a given predicate function/lambda returns true
/// Parameters:
///  - predicateFn: A function/lambda to test if the actionFn should be called. Return True to call the actionFn, False to continue waiting
///  - actionFn: A function/lambda to call when the predicateFn returns true. This will only be called once and then the timer will stop.
///  - checkInterval: How often, in seconds, to call predicateFn
hs.timer.waitUntil = function(predicateFn, actionFn, checkInterval) {
    if (typeof predicateFn !== 'function') {
        throw new Error("hs.timer.waitUntil(): predicate must be a function");
    }
    if (typeof actionFn !== 'function') {
        throw new Error("hs.timer.waitUntil(): action must be a function");
    }

    checkInterval = checkInterval || 1;

    const timer = hs.timer.create(checkInterval, function() {
        if (predicateFn()) {
            actionFn();
            timer.stop();
        }
    });

    return timer.start();
};

/// Wait to call a function/lambda until a given predicate function/lambda returns false
/// Parameters:
///  - predicateFn: A function/lambda to test if the actionFn should be called. Return False to call the actionFn, True to continue waiting
///  - actionFn: A function/lambda to call when the predicateFn returns False. This will only be called once and then the timer will stop.
///  - checkInterval: How often, in seconds, to call predicateFn
hs.timer.waitWhile = function(predicateFn, actionFn, checkInterval) {
    if (typeof predicateFn !== 'function') {
        throw new Error("hs.timer.waitWhile(): predicate must be a function");
    }
    if (typeof actionFn !== 'function') {
        throw new Error("hs.timer.waitWhile(): action must be a function");
    }

    checkInterval = checkInterval || 1;

    const timer = hs.timer.create(checkInterval, function() {
        if (!predicateFn()) {
            actionFn();
            timer.stop();
        }
    });

    return timer.start();
};

File: Hammerspoon 2/Modules/hs.timer/HSTimer.swift
//
//  TimerObject.swift
//  Hammerspoon 2
//
//  Created by Chris Jones on 06/11/2025.
//

import Foundation
import JavaScriptCore

/// Object representing a timer. You should not instantiate these yourself, but rather, use the methods in hs.timer to create them for you.
@objc protocol HSTimerAPI: HSTypeAPI, JSExport {
    /// The timer's interval in seconds
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// console.log(t.interval)
    /// ```
    @objc var interval: TimeInterval { get }

    /// Whether the timer repeats
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// console.log(t.repeats)
    /// ```
    @objc var repeats: Bool { get }

    /// Start the timer
    /// - Example:
    /// ```js
    /// const t = hs.timer.new(5, () => console.log("tick"), false)
    /// t.start()
    /// ```
    @objc func start()

    /// Stop the timer
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// t.stop()
    /// ```
    @objc func stop()

    /// Immediately fire the timer's callback
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => console.log("tick"))
    /// t.fire()
    /// ```
    @objc func fire()

    /// Check if the timer is currently running
    /// - Returns: true if the timer is running, false otherwise
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// console.log(t.running())
    /// ```
    @objc func running() -> Bool

    /// Get the number of seconds until the timer next fires
    /// - Returns: Seconds until next trigger, or a negative value if the timer is not running
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// console.log(t.nextTrigger())
    /// ```
    @objc func nextTrigger() -> TimeInterval

    /// Set when the timer should next fire
    /// - Parameter seconds: Number of seconds from now when the timer should fire
    /// - Example:
    /// ```js
    /// const t = hs.timer.doEvery(5, () => {})
    /// t.setNextTrigger(10)
    /// ```
    @objc func setNextTrigger(_ seconds: TimeInterval)
}

@_documentation(visibility: private)
@objc class HSTimer: NSObject, HSTimerAPI {
    @objc var typeName = "HSTimer"
    private var timer: Timer?
    private var callback: JSCallback?
    private let continueOnError: Bool

    @objc let interval: TimeInterval
    @objc let repeats: Bool

    init(interval: TimeInterval, repeats: Bool, callback: JSFunction, continueOnError: Bool = false) {
        self.interval = interval
        self.repeats = repeats
        self.continueOnError = continueOnError
        super.init()
        self.callback = JSCallback(value: callback, owner: self)
    }

    isolated deinit {
        destroy()
        AKDebug("HSTimer deinit")
    }

    func destroy() {
        stop()
        callback?.detach(from: self)
        callback = nil
    }

    @objc func start() {
        // If already running, don't create a new timer
        if timer?.isValid == true {
            return
        }

        timer = Timer.scheduledTimer(timeInterval: interval,
                                     target: self,
                                     selector: #selector(timerDidFire),
                                     userInfo: nil,
                                     repeats: repeats)

        // Add to common run loop modes so timer fires during modal dialogs, etc.
        if let timer {
            RunLoop.current.add(timer, forMode: .common)
        }
    }

    @objc func stop() {
        timer?.invalidate()
        timer = nil
    }

    @objc func fire() {
        // Fire immediately, bypassing the timer
        timerDidFire()
    }

    @objc func running() -> Bool {
        return timer?.isValid ?? false
    }

    @objc func nextTrigger() -> TimeInterval {
        guard let timer = timer, timer.isValid else {
            return -1
        }

        let fireDate = timer.fireDate
        let now = Date()
        return fireDate.timeIntervalSince(now)
    }

    @objc func setNextTrigger(_ seconds: TimeInterval) {
        guard let timer = timer, timer.isValid else {
            AKWarning("hs.timer:setNextTrigger(): Timer is not running")
            return
        }

        let newFireDate = Date(timeIntervalSinceNow: seconds)
        timer.fireDate = newFireDate
        return
    }

    @objc private func timerDidFire() {
        guard let callbackValue = callback?.value, callbackValue.isObject else {
            AKError("hs.timer: callback is not a function")
            if !continueOnError {
                stop()
            }
            return
        }

        // Call the callback. callSafely catches & logs any thrown JS exception
        // (incl. stack) tagged "hs.timer" and clears the engine exception slot;
        // it returns nil on throw so we can decide whether to stop the timer.
        let result = callbackValue.callSafely(withArguments: [], context: "hs.timer")
        if result == nil && !continueOnError {
            stop()
        }

        // For one-shot timers, clean up after firing
        if !repeats {
            stop()
        }
    }
}
